### PR TITLE
Added simple support for returning a multitarget loss

### PR DESCRIPTION
**Pull request by geohot**
Adds support for returning each target's loss as a metric. Couldn't find an easy way to do this with metrics without repeating all loss functions, and this is important to us for tuning our loss functions.

Feel free to reject or refactor, I know it's a feature and not a bugfix. Or maybe I'm just dumb and there's an easy way to do this (tried "loss" as the metric in the multi_output dictionary, didn't work, would have had to respecify each function by name)

### CODE REVIEW

**Comment by fchollet** [comment]
> Adds support for returning each target's loss as a metric. Couldn't find an easy way to do this with metrics without repeating all loss functions, and this is important to us for tuning our loss functions.

Currently the way you would do it is by passing everything you want to monitor to the `metrics` dict/list. E.g. if you are training with two outputs:

```python
model.compile(optimizer='rmsprop',
              loss={'out1': 'mse', 'out2': 'mae'},
              metrics={'out1': ['acc', 'mse'], 'out2': ['acc', 'mae']})
```

`metrics.py` contains mirror functions of every built-in loss function (note that the only difference between a loss function and a metric function is that the former returns one value per sample while the latter returns a single scalar).

Any reason why the above API doesn't work out for you?

Basically the motivation for the current API is: "loss" is just for optimization, "metrics" is what you want to monitor.

**Comment by geohot** [comment]
Thanks for the fast reply! Three comments on why we changed the code

1) We have custom loss functions and I didn't see a way to create custom metrics. The API doesn't seem to allow a function passed in instead of a string like "mse" since it goes through metrics_module.get.
2) There's potential for bugs in using different code for training loss evaluation. The loss functions shouldn't live in two different places.
3) Not sure Theano is smart enough to merge the two computes into one, though loss functions are usually so fast it's probably negligible.

**Comment by fchollet** [comment]
> We have custom loss functions and I didn't see a way to create custom metrics. The API doesn't seem to allow a function passed in instead of a string like "mse" since it goes through metrics_module.get.

Actually `module.get`, when passed a function rather than a string, will return that function. It's the same system used for objectives & optimizers. So you can pass a custom function as part of the `metrics` argument.

> There's potential for bugs in using different code for training loss evaluation. The loss functions shouldn't live in two different places.

That's a good point. A shortcut is to use the same code but wrap your loss function into a `K.mean()` to turn it into a metric:

```python
model.compile('rmsprop', loss=my_loss, metrics=[lambda x, y: K.mean(my_loss(x, y))])
```
(in this case the metric would just be displayed as "lambda", to display its name just wrap it into a named function -- okay maybe it's not the most practical thing).

> Not sure Theano is smart enough to merge the two computes into one, though loss functions are usually so fast it's probably negligible.

Theano definitely does similar optimizations, so maybe, but I wouldn't bet on it. In any case, yes, I wouldn't worry about the computation cost of a loss function (at least for all those that built-in).

Remains that it would indeed make sense to add all separate loss functions to `metrics`. Let me think about this a bit.

**Comment by geohot** [comment]
Oh my bad, I didn't actually just try passing in a function, I thought I read the code. Are there docs on metrics?

There's also the loss function scaling to replicate if using metrics. And perhaps there's a way for metrics to be refactored to not copy the code of all the loss functions in metrics.py

**Comment by EderSantana** [comment]
as far as my counting goes, this is actually the 3rd time we get a PR about using custom cost functions (myself included when I started with Keras). +1 for to an explicit custom cost function example in the docs.

**Comment by fchollet** [comment]
>  +1 to an explicit custom cost function example in the docs.

Yes, I also think we need an example script with 1) custom loss functions and 2) multiple loss functions at the same time. Especially since there are some unique issues to consider when using multiple losses (loss scaling, for one).

@geohot let's do this. A few points:
- I don't think we need a new keyword argument. We can just look at `self.outputs`: if there are multiple outputs, then we will name the main loss "total_loss" in the `metrics` list, and we will add add each separate loss to the `metrics` list under the name `output_name + '_loss'`
- I don't think we should log the post-scaling losses (after `loss_weight` has been applied), rather the user is going to care about the raw loss value. So we should add to `metrics` the output of `weighted_loss(y_true, y_pred, sample_weight, mask)`.

**Comment by rpinsler** [comment]
Another common requirement is to provide additional arguments to the cost function, see e.g. #2121. Maybe an example would be helpful, too.

**Comment by fchollet** [comment]
@geohot do you want to update the PR with the changes, or should I do it?

**Comment by EderSantana** [comment]
> That's a good point. A shortcut is to use the same code but wrap your loss function into a K.mean() to turn it into a metric:

@fchollet what if we import `objectives` inside `metrics`? Won't its `get` see all the objectives without need to repeat?

**Comment by fchollet** [comment]
Merged along with the changes discussed above in https://github.com/fchollet/keras/pull/2504#issuecomment-214607808
